python/extensions/python.bzl

```diff
@@ -60,82 +60,97 @@ def _python_register_toolchains(toolchain_attr, version_constraint):
     )
 
 def _python_impl(module_ctx):
-    # Use to store all of the toolchains
+    # The toolchain info structs to register, in the order to register them in.
     toolchains = []
 
-    # Used to check if toolchains already exist
-    toolchain_names = []
-
-    # Used to store toolchains that are in sub modules.
-    sub_toolchains_map = {}
+    # We store the default toolchain separately to ensure it is the last
+    # toolchain added to toolchains.
     default_toolchain = None
-    python_versions = {}
+
+    # Map of toolchain name to registering module
+    global_toolchain_names = {}
+
+    # Map of string Major.Minor to the toolchain name and module name
+    global_toolchain_versions = {}
 
     for mod in module_ctx.modules:
+        module_toolchain_names = []
+        module_toolchain_versions = []
+
         for toolchain_attr in mod.tags.toolchain:
-            # If we are in the root module we always register the toolchain.
-            # We wait to register the default toolchain till the end.
+            toolchain_name = toolchain_attr.name
+
+            # Duplicate names within a module indicate a misconfigured module.
+            if toolchain_name in module_toolchain_names:
+                _fail_duplicate_module_toolchain_name(mod.name, toolchain_name)
+            module_toolchain_names.append(toolchain_name)
+
+            # Ignore name collisions in the global scope because there isn't
+            # much else that can be done. Modules don't know and can't control
+            # what other modules do, so the first in the dependency graph wins.
+            if toolchain_name in global_toolchain_names:
+                _warn_duplicate_global_toolchain_name(
+                    toolchain_name,
+                    first_module = global_toolchain_names[toolchain_name],
+                    second_module = mod.name,
+                )
+                continue
+            global_toolchain_names[toolchain_name] = mod.name
+
+            # Duplicate versions within a module indicate a misconfigured module.
+            toolchain_version = toolchain_attr.python_version
+            if toolchain_version in module_toolchain_versions:
+                _fail_duplicate_module_toolchain_version(toolchain_version, mod.name)
+            module_toolchain_versions.append(toolchain_version)
+
+            # Ignore version collisions in the global scope because there isn't
+            # much else that can be done. Modules don't know and can't control
+            # what other modules do, so the first in the dependency graph wins.
+            if toolchain_version in global_toolchain_versions:
```

Comment:
Should we fail if we have a root module toolchain with the same version?  They can fix that.

Reply:
Yes, that's handled around L103, where, within a module, it's checked that a version hasn't already been requested.